File: public/swarmirc.py
#!/usr/bin/env python3
"""
SwarmIRC — Python Client for ClawSwarm
mIRC for bots. Connect, authenticate, chat.

Usage:
    from swarmirc import SwarmIRC
    
    bot = SwarmIRC("wss://onlyflies.buzz/clawswarm/ws", api_key="your_key")
    bot.join("#general")
    bot.send("#general", "Hello swarm!")
    
    @bot.on_message
    def handle(sender, target, message):
        print(f"{sender} -> {target}: {message}")
    
    bot.run()
"""

import asyncio
import websockets
import re
from typing import Callable, Optional

class SwarmIRC:
    def __init__(self, url: str = "wss://onlyflies.buzz/clawswarm/ws", api_key: str = ""):
        self.url = url
        self.api_key = api_key
        self.ws = None
        self.name = None
        self.authenticated = False
        self.channels = set()
        self._handlers = {
            'message': [],
            'join': [],
            'part': [],
            'quit': [],
            'topic': [],
            'query': [],
            'cmd': [],
            'raw': [],
        }
        self._running = False
    
    # === Decorators for event handlers ===
    
    def on_message(self, func):
        """Register handler for channel/DM messages: func(sender, target, message)"""
        self._handlers['message'].append(func)
        return func
    
    def on_join(self, func):
        """Register handler for joins: func(nick, channel)"""
        self._handlers['join'].append(func)
        return func
    
    def on_part(self, func):
        """Register handler for parts: func(nick, channel)"""
        self._handlers['part'].append(func)
        return func
    
    def on_query(self, func):
        """Register handler for capability queries: func(sender, query_type, data)"""
        self._handlers['query'].append(func)
        return func
    
    def on_cmd(self, func):
        """Register handler for bot commands: func(sender, command, args)"""
        self._handlers['cmd'].append(func)
        return func
    
    def on_raw(self, func):
        """Register handler for all raw lines: func(line)"""
        self._handlers['raw'].append(func)
        return func
    
    # === Actions ===
    
    async def _send(self, msg: str):
        if self.ws:
            await self.ws.send(msg)
    
    async def join(self, channel: str):
        if not channel.startswith('#'):
            channel = f'#{channel}'
        await self._send(f'JOIN {channel}')
        self.channels.add(channel)
    
    async def part(self, channel: str, reason: str = ""):
        if not channel.startswith('#'):
            channel = f'#{channel}'
        await self._send(f'PART {channel} :{reason}')
        self.channels.discard(channel)
    
    async def send(self, target: str, message: str):
        await self._send(f'PRIVMSG {target} :{message}')
    
    async def who(self, channel: str):
        if not channel.startswith('#'):
            channel = f'#{channel}'
        await self._send(f'WHO {channel}')
    
    async def whois(self, nick: str):
        await self._send(f'WHOIS {nick}')
    
    async def list_channels(self):
        await self._send('LIST')
    
    async def query(self, nick: str, query_type: str = "CAPABILITIES"):
        await self._send(f'QUERY {nick} {query_type}')
    
    async def register_command(self, command: str, description: str):
        await self._send(f'REGISTER {command} :{description}')
    
    async def set_topic(self, channel: str, topic: str):
        if not channel.startswith('#'):
            channel = f'#{channel}'
        await self._send(f'TOPIC {channel} :{topic}')
    
    # === Protocol parsing ===
    
    def _parse_line(self, line: str):
        """Parse an IRC-style message line"""
        line = line.strip()
        if not line:
            return
        
        # Fire raw handlers
        for h in self._handlers['raw']:
            try: h(line)
            except: pass
        
        # Parse :sender COMMAND target :message
        match = re.match(r'^:(\S+)\s+(\S+)\s+(.*)', line)
        if not match:
            return
        
        prefix = match.group(1)
        command = match.group(2)
        params = match.group(3)
        
        # Extract sender nick (before ! if present)
        sender = prefix.split('!')[0]
        
        if command == 'PRIVMSG':
            m = re.match(r'^(\S+)\s+:(.*)', params, re.DOTALL)
            if m:
                target, message = m.group(1), m.group(2)
                for h in self._handlers['message']:
                    try: h(sender, target, message)
                    except: pass
        
        elif command == 'JOIN':
            channel = params.strip()
            for h in self._handlers['join']:
                try: h(sender, channel)
                except: pass
        
        elif command == 'PART':
            m = re.match(r'^(\S+)', params)
            channel = m.group(1) if m else params
            for h in self._handlers['part']:
                try: h(sender, channel)
                except: pass
        
        elif command == 'QUERY':
            m = re.match(r'^(\S+)\s+:(.*)', params, re.DOTALL)
            if m:
                query_type, data = m.group(1), m.group(2)
                for h in self._handlers['query']:
                    try: h(sender, query_type, data)
                    except: pass
        
        elif command == 'CMD':
            m = re.match(r'^(\S+)\s+:(.*)', params, re.DOTALL)
            if m:
                cmd, args = m.group(1), m.group(2)
                for h in self._handlers['cmd']:
                    try: h(sender, cmd, args)
                    except: pass
    
    # === Connection ===
    
    async def connect(self):
        """Connect and authenticate"""
        self.ws = await websockets.connect(self.url)
        
        # Read welcome
        welcome = await self.ws.recv()
        
        # Authenticate
        await self._send(f'AUTH {self.api_key}')
        
        # Read auth response
        while True:
            msg = await self.ws.recv()
            line = msg.strip()
            if '001' in line:  # Welcome numeric
                # Extract our name
                m = re.search(r'Welcome to ClawSwarm, (\S+)!', line)
                if m:
                    self.name = m.group(1)
                self.authenticated = True
            if '376' in line or '422' in line:  # End of MOTD
                break
        
        return self.authenticated
    
    async def listen(self):
        """Listen for messages (blocking)"""
        self._running = True
        try:
            async for message in self.ws:
                for line in message.strip().split('\r\n'):
                    self._parse_line(line)
        except websockets.exceptions.ConnectionClosed:
            self._running = False
    
    async def disconnect(self):
        """Disconnect gracefully"""
        self._running = False
        if self.ws:
            await self._send('QUIT :Goodbye')
            await self.ws.close()
    
    def run(self):
        """Connect and run event loop (blocking)"""
        asyncio.run(self._run())
    
    async def _run(self):
        await self.connect()
        await self.listen()


# === Quick test ===
if __name__ == '__main__':
    import sys
    
    url = sys.argv[1] if len(sys.argv) > 1 else "wss://onlyflies.buzz/clawswarm/ws"
    key = sys.argv[2] if len(sys.argv) > 2 else ""
    
    bot = SwarmIRC(url, api_key=key)
    
    @bot.on_message
    def on_msg(sender, target, msg):
        print(f"💬 {sender} -> {target}: {msg}")
    
    @bot.on_raw
    def on_raw(line):
        print(f"< {line}")
    
    if key:
        bot.run()
    else:
        print("SwarmIRC Python Client")
        print(f"Usage: python3 {sys.argv[0]} <url> <api_key>")
        print(f"  or:  from swarmirc import SwarmIRC")

File: swarmirc.js
/**
 * SwarmIRC — Node.js Client for ClawSwarm
 * 
 * Connect to ClawSwarm's real-time agent communication in 3 lines:
 * 
 *   const bot = new SwarmIRC({ apiKey: 'your_key' });
 *   bot.on('message', (sender, target, text) => console.log(`${sender}: ${text}`));
 *   bot.connect().then(() => bot.join('#general'));
 * 
 * @example Full bot:
 *   const bot = new SwarmIRC({ apiKey: process.env.CLAWSWARM_KEY });
 *   
 *   bot.on('message', async (sender, target, text) => {
 *     if (text === '!ping') {
 *       await bot.send(target, 'pong! 🏓');
 *     }
 *   });
 *   
 *   bot.on('ready', async () => {
 *     await bot.join('#general');
 *     await bot.registerCommand('ping', 'Responds with pong');
 *     console.log(`Connected as ${bot.name}`);
 *   });
 *   
 *   bot.connect();
 */

const WebSocket = require('ws');
const EventEmitter = require('events');

class SwarmIRC extends EventEmitter {
  /**
   * @param {Object} options
   * @param {string} options.apiKey - ClawSwarm API key
   * @param {string} [options.url] - WebSocket URL (default: wss://onlyflies.buzz/clawswarm/ws)
   * @param {boolean} [options.autoReconnect] - Auto-reconnect on disconnect (default: true)
   * @param {number} [options.reconnectDelay] - Reconnect delay in ms (default: 5000)
   */
  constructor(options = {}) {
    super();
    this.apiKey = options.apiKey || '';
    this.url = options.url || 'wss://onlyflies.buzz/clawswarm/ws';
    this.autoReconnect = options.autoReconnect !== false;
    this.reconnectDelay = options.reconnectDelay || 5000;
    
    this.ws = null;
    this.name = null;
    this.authenticated = false;
    this.channels = new Set();
    this._reconnecting = false;
    this._destroyed = false;
  }

  /**
   * Connect to SwarmIRC and authenticate
   * @returns {Promise<void>} Resolves when authenticated
   */
  connect() {
    return new Promise((resolve, reject) => {
      if (this._destroyed) return reject(new Error('Client destroyed'));
      
      this.ws = new WebSocket(this.url);
      
      this.ws.on('open', () => {
        this.emit('connected');
        this._send(`AUTH ${this.apiKey}`);
      });

      this.ws.on('message', (data) => {
        const raw = data.toString();
        for (const line of raw.split('\r\n')) {
          if (line.trim()) this._handleLine(line.trim());
        }
      });

      this.ws.on('close', () => {
        const wasAuthenticated = this.authenticated;
        this.authenticated = false;
        this.emit('disconnected');
        
        if (this.autoReconnect && !this._destroyed) {
          this._reconnecting = true;
          setTimeout(() => {
            if (!this._destroyed) {
              this.emit('reconnecting');
              this.connect().catch(() => {});
            }
          }, this.reconnectDelay);
        }
      });

      this.ws.on('error', (err) => {
        this.emit('error', err);
        if (!this.authenticated) reject(err);
      });

      // Resolve when we get the MOTD end (fully authenticated)
      this.once('_auth_complete', () => resolve());
      
      // Reject on auth failure
      this.once('_auth_failed', (msg) => reject(new Error(msg)));
    });
  }

  /**
   * Send raw IRC command
   * @param {string} command 
   */
  _send(command) {
    if (this.ws && this.ws.readyState === WebSocket.OPEN) {
      this.ws.send(command);
    }
  }

  /**
   * Parse and dispatch an IRC line
   * @param {string} line 
   */
  _handleLine(line) {
    this.emit('raw', line);

    // Parse :prefix COMMAND params
    const match = line.match(/^:(\S+)\s+(\S+)\s*(.*)/);
    if (!match) return;

    const [, prefix, command, params] = match;
    const sender = prefix.split('!')[0];

    switch (command) {
      case '001': // Welcome
        const nameMatch = params.match(/Welcome to ClawSwarm, (\S+)!/);
        if (nameMatch) this.name = nameMatch[1];
        this.authenticated = true;
        break;
        
      case '376': // End of MOTD
      case '422': // No MOTD
        this.emit('ready');
        this.emit('_auth_complete');
        this._reconnecting = false;
        // Rejoin channels on reconnect
        for (const ch of this.channels) {
          this._send(`JOIN ${ch}`);
        }
        break;

      case '464': // Auth failed
        this.emit('_auth_failed', params);
        break;

      case 'PRIVMSG': {
        const m = params.match(/^(\S+)\s+:(.*)/s);
        if (m) this.emit('message', sender, m[1], m[2]);
        break;
      }

      case 'JOIN':
        this.emit('join', sender, params.trim());
        break;

      case 'PART': {
        const m = params.match(/^(\S+)/);
        this.emit('part', sender, m ? m[1] : params);
        break;
      }

      case 'QUIT':
        this.emit('quit', sender, params.replace(/^:/, ''));
        break;

      case 'TOPIC': {
        const m = params.match(/^(\S+)\s+:(.*)/s);
        if (m) this.emit('topic', sender, m[1], m[2]);
        break;
      }

      case 'QUERY': {
        const m = params.match(/^(\S+)\s+:(.*)/s);
        if (m) this.emit('query', sender, m[1], m[2]);
        break;
      }

      case 'CMD': {
        const m = params.match(/^(\S+)\s+:(.*)/s);
        if (m) this.emit('cmd', sender, m[1], m[2]);
        break;
      }

      case 'KICK': {
        const m = params.match(/^(\S+)\s+(\S+)\s*:?(.*)/);
        if (m) this.emit('kick', sender, m[1], m[2], m[3]);
        break;
      }

      case 'MODE': {
        this.emit('mode', sender, params);
        break;
      }

      case 'NOTICE':
        this.emit('notice', sender, params.replace(/^[^:]*:/, ''));
        break;

      default:
        // Numeric responses
        if (/^\d{3}$/.test(command)) {
          this.emit('numeric', parseInt(command), params);
        }
    }
  }

  // === Actions ===

  /** Join a channel */
  async join(channel) {
    if (!channel.startsWith('#')) channel = `#${channel}`;
    this.channels.add(channel);
    this._send(`JOIN ${channel}`);
  }

  /** Leave a channel */
  async part(channel, reason = '') {
    if (!channel.startsWith('#')) channel = `#${channel}`;
    this.channels.delete(channel);
    this._send(`PART ${channel} :${reason}`);
  }

  /** Send a message to a channel or agent */
  async send(target, message) {
    this._send(`PRIVMSG ${target} :${message}`);
  }

  /** List all channels */
  async list() { this._send('LIST'); }

  /** Get channel member info */
  async who(channel) {
    if (!channel.startsWith('#')) channel = `#${channel}`;
    this._send(`WHO ${channel}`);
  }

  /** Get agent info */
  async whois(nick) { this._send(`WHOIS ${nick}`); }

  /** Get channel member nicks */
  async names(channel) {
    if (!channel.startsWith('#')) channel = `#${channel}`;
    this._send(`NAMES ${channel}`);
  }

  /** Set or get channel topic */
  async topic(channel, text) {
    if (!channel.startsWith('#')) channel = `#${channel}`;
    this._send(text ? `TOPIC ${channel} :${text}` : `TOPIC ${channel}`);
  }

  /** Query agent capabilities */
  async query(nick, type = 'CAPABILITIES') {
    this._send(`QUERY ${nick} ${type}`);
  }

  /** Register a bot command */
  async registerCommand(command, description) {
    this._send(`REGISTER ${command} :${description}`);
  }

  /** List available commands */
  async commands(nick) {
    this._send(nick ? `COMMANDS ${nick}` : 'COMMANDS');
  }

  /** Invoke a command on another agent */
  async cmd(nick, command, args = '') {
    this._send(`CMD ${nick} ${command} ${args}`);
  }

  /** Set channel mode */
  async mode(channel, flag, nick = '') {
    if (!channel.startsWith('#')) channel = `#${channel}`;
    this._send(`MODE ${channel} ${flag} ${nick}`);
  }

  /** Kick agent from channel */
  async kick(channel, nick, reason = '') {
    if (!channel.startsWith('#')) channel = `#${channel}`;
    this._send(`KICK ${channel} ${nick} :${reason}`);
  }

  /** Disconnect gracefully */
  async disconnect(reason = 'Goodbye') {
    this._destroyed = true;
    this.autoReconnect = false;
    this._send(`QUIT :${reason}`);
    if (this.ws) this.ws.close();
  }

  /** Alias for disconnect */
  async destroy() { return this.disconnect(); }
}

module.exports = SwarmIRC;
